### PR TITLE
Created a Python 3 module for simpler building of Python extension mo…

### PR DESCRIPTION
**Pull request by jpakkane**
…dules.

### CODE REVIEW

File: mesonbuild/environment.py [hunk 1]
```diff
@@ -853,3 +853,11 @@ def need_exe_wrapper(self):
             # host and build systems are the same
             return False
         return True
+
+
+class MachineInfo:
```
Comment (QuLogic) [inline mesonbuild/environment.py]: `namedtuple`?

**Comment by jpakkane** [comment]
FTR I really don't like what this does to the module API (if we can even be said to have such a thing). We definitely need something better.

**Comment by nirbheek** [comment]
What specifically do you not like?

**Comment by jpakkane** [comment]
Thus far modules have created objects by calling directly into build etc. This one does it by calling into interpreter methods. Having two different ways of doing roughly the same thing is confusing sine you need to "know" which one to do. Also need to be careful not to, for example, add created targets twice, once when calling the interpreter method and once when returning the same value from module method call.

**Comment by nirbheek** [comment]
I think this can be reimplemented with the new module support and it will be all perfectly fine.

**Comment by ignatenkobrain** [comment]
Would be nice to have "generic" python module, so user can specify whether he wants py3 or py3. And would be very good if it would be possible to do both at the same time. Building whole project second time is not always fast.

File: mesonbuild/interpreter.py [hunk 2]
```diff
@@ -354,23 +354,27 @@ class BuildMachine(InterpreterObject):
...
-        self.methods.update({'system': self.system_method,
-                             'cpu_family': self.cpu_family_method,
-                             'cpu': self.cpu_method,
-                             'endian': self.endian_method,
-                             })
+        self.held_object = environment.MachineInfo(environment.detect_system(),
+                                                   environment.detect_cpu_family(self.compilers),
+                                                   environment.detect_cpu(self.compilers),
+                                                   sys.byteorder)
+        self.methods.update({'system' : self.system_method,
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[flake8]__
* [E203] whitespace before ':'

File: mesonbuild/interpreter.py [hunk 3]
```diff
@@ -354,23 +354,27 @@ class BuildMachine(InterpreterObject):
...
+                             'cpu_family' : self.cpu_family_method,
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[flake8]__
* [E203] whitespace before ':'

File: mesonbuild/interpreter.py [hunk 4]
```diff
@@ -354,23 +354,27 @@ class BuildMachine(InterpreterObject):
...
+                             'cpu' : self.cpu_method,
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[flake8]__
* [E203] whitespace before ':'

File: mesonbuild/interpreter.py [hunk 5]
```diff
@@ -354,23 +354,27 @@ class BuildMachine(InterpreterObject):
...
+                             'endian' : self.endian_method,
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[flake8]__
* [E203] whitespace before ':'

File: mesonbuild/interpreter.py [hunk 6]
```diff
@@ -354,23 +354,27 @@ class BuildMachine(InterpreterObject):
...
+                            })
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[flake8]__
* [E124] closing bracket does not match visual indentation

File: mesonbuild/interpreter.py [hunk 7]
```diff
@@ -383,23 +387,27 @@ def __init__(self, cross_info):
...
-        self.methods.update({'system': self.system_method,
-                             'cpu': self.cpu_method,
-                             'cpu_family': self.cpu_family_method,
-                             'endian': self.endian_method,
-                             })
+        self.held_object = environment.MachineInfo(cross_info['system'],
+                                                   cross_info['cpu_family'],
+                                                   cross_info['cpu'],
+                                                   cross_info['endian'])
+        self.methods.update({'system' : self.system_method,
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[flake8]__
* [E203] whitespace before ':'

File: mesonbuild/interpreter.py [hunk 8]
```diff
@@ -383,23 +387,27 @@ def __init__(self, cross_info):
...
+                             'cpu' : self.cpu_method,
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[flake8]__
* [E203] whitespace before ':'

File: mesonbuild/interpreter.py [hunk 9]
```diff
@@ -383,23 +387,27 @@ def __init__(self, cross_info):
...
+                             'cpu_family' : self.cpu_family_method,
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[flake8]__
* [E203] whitespace before ':'

File: mesonbuild/interpreter.py [hunk 10]
```diff
@@ -383,23 +387,27 @@ def __init__(self, cross_info):
...
+                             'endian' : self.endian_method,
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[flake8]__
* [E203] whitespace before ':'

File: mesonbuild/interpreter.py [hunk 11]
```diff
@@ -383,23 +387,27 @@ def __init__(self, cross_info):
...
+                            })
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[flake8]__
* [E124] closing bracket does not match visual indentation

File: mesonbuild/interpreter.py [hunk 12]
```diff
@@ -354,23 +354,27 @@ class BuildMachine(InterpreterObject):
...
+        self.held_object = environment.MachineInfo(environment.detect_system(),
+                                                   environment.detect_cpu_family(self.compilers),
+                                                   environment.detect_cpu(self.compilers),
+                                                   sys.byteorder)
         self.methods.update({'system': self.system_method,
                              'cpu_family': self.cpu_family_method,
                              'cpu': self.cpu_method,
                              'endian': self.endian_method,
-                             })
+                            })
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[flake8]__
* [E124] closing bracket does not match visual indentation

File: mesonbuild/interpreter.py [hunk 13]
```diff
@@ -383,23 +387,27 @@ def __init__(self, cross_info):
...
+        self.held_object = environment.MachineInfo(cross_info['system'],
+                                                   cross_info['cpu_family'],
+                                                   cross_info['cpu'],
+                                                   cross_info['endian'])
         self.methods.update({'system': self.system_method,
                              'cpu': self.cpu_method,
                              'cpu_family': self.cpu_family_method,
                              'endian': self.endian_method,
-                             })
+                            })
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[flake8]__
* [E124] closing bracket does not match visual indentation

**Pull request merged**